src/middlewared/middlewared/migration/0009_nvram_attr_vms.py

```diff
@@ -0,0 +1,67 @@
+import contextlib
+import os
+import shutil
+
+from middlewared.plugins.zfs_.utils import zvol_path_to_name
+
+
+DEFAULT_NVRAM_FOLDER_PATH = '/var/lib/libvirt/qemu/nvram'
+LIBVIRT_QEMU_UID = 64055
+LIBVIRT_QEMU_GID = 64055
+
+
+def get_root_ds_from_disk_device(disk_device):
+    with contextlib.suppress(ValueError):
+        return zvol_path_to_name(disk_device['attributes']['path']).split('/')[0]
+
+
+def migrate(middleware):
+    existing_nvram_files = os.listdir(DEFAULT_NVRAM_FOLDER_PATH)
+    existing_root_datasets = [
+        ds['id'] for ds in middleware.call_sync(
+            'pool.dataset.query', [], {'extra': {'properties': ['name'], 'retrieve_children': False}}
+        )
+    ]
+    default_ds_to_use = existing_root_datasets[0] if existing_root_datasets else None
+    for vm in middleware.call_sync('vm.query', [['bootloader', '=', 'UEFI'], ['nvram_location', '=', None]]):
+        try:
+            migrate_vm_nvram_file(middleware, existing_nvram_files, vm, default_ds_to_use, existing_root_datasets)
+        except Exception:
+            middleware.logger.error('Failed to migrate nvram file for VM %r(%r)', vm['name'], vm['id'], exc_info=True)
+
+
+def migrate_vm_nvram_file(middleware, existing_nvram_files, vm, default_ds_to_use, existing_root_datasets):
+    file_name = f'{vm["id"]}_{vm["name"]}_VARS.fd'
+    root_ds_to_use = default_ds_to_use
+    for disk_device in middleware.call_sync('vm.device.query', [['vm', '=', vm['id']], ['dtype', '=', 'DISK']]):
+        if (root_ds := get_root_ds_from_disk_device(disk_device)) and root_ds in existing_root_datasets:
+            root_ds_to_use = root_ds
+            break
+
+    # We will copy the file here if it exists and update the nvram_location
+    if root_ds_to_use:
+        new_path = os.path.join('/mnt', root_ds_to_use, file_name)
+        i = 2
+        while True:
+            if not os.path.exists(new_path):
+                break
+            new_path = os.path.join('/mnt', root_ds_to_use, f'{file_name}_{i}')
```

Comment:
Will the files be put directly to the root dataset? Can we consider putting them along the disk files/zvols?

Reply:
@themylogin i thought there would be more edge cases which can arise like if zvol is `pool/dataset/zvol` and `pool/dataset` has readonly property set so it can cause problems with the migration.
Should I go ahead with trying this anyways or perhaps add a fallback to root dataset in that case?

Reply:
@yocalebo need your opinion here, personally I think we should keep similar files close, that's the strategy most users choose. Falling back to root dataset is a great idea.

Reply:
As much as I _dont_ like saying it, I believe we should put these UEFI files on the system dataset. We already have all he logic in place to manage it.